api/handler/v1beta1/job.go

```diff
@@ -13,46 +14,89 @@ import (
 	"google.golang.org/grpc/status"
 )
 
-func (sv *RuntimeServiceServer) DeployJobSpecification(req *pb.DeployJobSpecificationRequest, respStream pb.RuntimeService_DeployJobSpecificationServer) error {
+func (sv *RuntimeServiceServer) DeployJobSpecification(stream pb.RuntimeService_DeployJobSpecificationServer) error {
 	startTime := time.Now()
-
-	namespaceSpec, err := sv.namespaceService.Get(respStream.Context(), req.GetProjectName(), req.GetNamespaceName())
-	if err != nil {
-		return mapToGRPCErr(sv.l, err, "unable to get namespace")
-	}
-
-	var jobsToKeep []models.JobSpec
-	for _, reqJob := range req.GetJobs() {
-		adaptJob, err := sv.adapter.FromJobProto(reqJob)
+	for {
+		req, err := stream.Recv()
 		if err != nil {
-			return status.Errorf(codes.Internal, "%s: cannot adapt job %s", err.Error(), reqJob.GetName())
+			if err == io.EOF {
+				break
+			}
+			stream.Send(&pb.DeployJobSpecificationResponse{
+				Success: false,
+				Ack:     true,
+				Message: err.Error(),
+			})
+			return err
 		}
-
-		err = sv.jobSvc.Create(respStream.Context(), namespaceSpec, adaptJob)
+		namespaceSpec, err := sv.namespaceService.Get(stream.Context(), req.GetProjectName(), req.GetNamespaceName())
 		if err != nil {
-			return status.Errorf(codes.Internal, "%s: failed to save %s", err.Error(), adaptJob.Name)
+			stream.Send(&pb.DeployJobSpecificationResponse{
+				Success: false,
+				Ack:     true,
+				Message: err.Error(),
+			})
+			return mapToGRPCErr(sv.l, err, "unable to get namespace")
 		}
-		jobsToKeep = append(jobsToKeep, adaptJob)
-	}
 
-	observers := new(progress.ObserverChain)
-	observers.Join(sv.progressObserver)
-	observers.Join(&jobSyncObserver{
-		stream: respStream,
-		log:    sv.l,
-		mu:     new(sync.Mutex),
-	})
+		var jobsToKeep []models.JobSpec
+		for _, reqJob := range req.GetJobs() {
```

Comment:
We are sending a list of jobs on a bi-directional stream, is this intentional ?

Reply:
Yes, as of now, we send list of jobs per one request. We utilize the bidirectional stream for namespaces level instead of jobs